src/lib.rs

```diff
@@ -0,0 +1,191 @@
+// Copyright 2019 King's College London.
+// Created by the Software Development Team <http://soft-dev.org/>.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+#![feature(yk_swt)]
+#![feature(test)]
+
+extern crate test;
+
+use core::yk_swt::MirLoc;
+use std::iter::{IntoIterator, Iterator};
+
+mod swt;
+
+/// Generic representation of a trace of MIR block locations.
+pub trait MirTrace {
+    /// Returns the length of the trace, measured in MIR locations.
+    fn len(&self) -> usize;
+    /// Returns the MIR location and index `idx`.
+    fn loc(&self, idx: usize) -> &MirLoc;
+}
+
+/// An iterator over a MIR trace.
+pub struct MirTraceIntoIter<'a> {
+    trace: &'a MirTrace,
+    next_idx: usize
+}
+
+impl<'a> IntoIterator for &'a dyn MirTrace {
+    type Item = &'a MirLoc;
+    type IntoIter = MirTraceIntoIter<'a>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        MirTraceIntoIter {
+            trace: self,
+            next_idx: 0
+        }
+    }
+}
+
+impl<'a> Iterator for MirTraceIntoIter<'a> {
+    type Item = &'a MirLoc;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.next_idx < self.trace.len() {
+            let ret = self.trace.loc(self.next_idx);
+            self.next_idx += 1;
+            Some(ret)
+        } else {
+            None // No more locations.
+        }
+    }
+}
+
+/// The different ways by which we can collect a trace.
+#[derive(Clone, Copy)]
+pub enum TracingKind {
+    /// Software tracing via ykrustc.
+    SoftwareTracing,
+    /// Hardware tracing via ykrustc + hwtracer.
+    HardwareTracing
+}
+
+/// Represents a thread which is currently tracing.
+pub trait ThreadTracer {
+    /// Stops tracing on the current thread, returning the trace on success. `None` is returned on
+    /// error or if the trace was invalidated.
+    fn stop_tracing(self: Box<Self>) -> Option<Box<dyn MirTrace>>;
+}
+
+/// Start tracing on the current thread using the specified tracing kind.
+/// A thread tracer is returned through which tracing can be stopped and the trace obtained. The
+/// user must ensure that only one thread tracer exists at any one time per-thread, otherwise
+/// undefined behaviour will result.
+pub fn start_tracing(kind: TracingKind) -> Box<dyn ThreadTracer> {
```

Comment:
Is there a function somewhere which says "start tracing using whatever the best available tracer is"?

Reply:
I've not implemented one, no. We could have `kind` an `Option` where `None` means "just frigging choose one, will ya!".